### DISCUSSION TITLE
Private properties problem

### DISCUSSION THREAD

**Discussion by notmedia** [mikro-orm/mikro-orm#1941] [Q&A] [answered]
Hi, I have an entity like this:

```ts
@Entity({ tableName: 'my_entities' })
export class MyEntity {
  [EntityRepositoryType]?: MyEntityRepository;

  @PrimaryKey({ type: 'uuid', defaultRaw: 'uuid_generate_v4()' })
  private _uuid!: string;

  get uuid() {
    return this._uuid;
  }

  @Property({ fieldName: 'user_uuid' })
  private _userUuid!: string;

  get userUuid() {
    return this._userUuid;
  }
```

When I query this entity with userUuid field  I get an error:

```ts
this.repository.find({ userUuid: '12312' });

Error: Trying to query by not existing property MyEntity.userUuid
```

I tried to pass property name, but it doesn't work:
```ts
  @Property({ name: 'userUuid', fieldName: 'user_uuid' })
  private _userUuid!: string;
```

How I can achieve this with private props?

**Comment by B4nan** [comment, marked as answer]
This is not possible, there is no way to infer private property keys in TS. You need to use public properties or type cast/ts-ignore.

Some discussion about this problem is here: https://github.com/microsoft/TypeScript/issues/22677

**Reply by notmedia** [reply]
Thank you for fast reply.

I just console-debugged this in 
![image](https://user-images.githubusercontent.com/12993018/122576663-9592ec00-d05a-11eb-8804-64652046ebc0.png)

and the output for _userUuid is

```ts
 {
    name: '_userUuid',
    reference: 'scalar',
    getter: false,
    setter: false,
    type: 'string',
    fieldNames: [ 'user_uuid' ],
    columnTypes: [ 'varchar(255)' ],
    unsigned: undefined
  },
```

And this check wont pass `this.prop = meta.props.find(prop => prop.name === k || (prop.fieldNames || []).includes(k));` because the `name` is with underscore at start. Maybe if I can override this name with @Property({ name }) it would work. Or implement custom naming strategy?

**Reply by notmedia** [reply]
And another question) I use private properties to locking this props for modifying from outside. Maybe you can advice another approach?

**Reply by B4nan** [reply]
I really don't think what you are trying to do is possible. Can't help much, I am using public properties myself, as otherwise TS would not work and all the strict types would be useless. What you can try to do is to annotate the getter instead of the private property.

**Reply by B4nan** [reply]
Maybe you could use `EntitySchema` instead of decorators and use both class with private props and interface with public ones, but don't have any examples for that, just an idea.

**Reply by notmedia** [reply]
Thanks, I chose the decorators approach because I use another libraries which helps me. Like automapper, so it's preferred option instead of EntitySchema

And my idea, I'll try to explain :) 

What if we apply `propertyName` which mikroorm need to use in meta with Property decorator.

For example, 

```ts
  @Property({ propertyName: 'userUuid', fieldName: 'user_uuid' })
  private _userUuid!: string;
```

After that in CriteriaNode we can try to find prop by this `propertyName` meta field. Like it do with `fieldName` prop?

If we can't do this, I switch to public properties. But I prefer to use private props in entity models, this is more controllable and programmers make bugs less

**Reply by B4nan** [reply]
The problem is that the same property name would be used for writing/hydrating, so you would still have to define a setter too, otherwise the hydration would fail. 

Feel free to give it a try and craft a PR, if you can make it work, will be happy to accept it. I feel like it would have to be quite fundamental refactoring to how things work internally if we would like to not have the setter and hydrate to the private property directly (as there will be probably lots of places where this would have to be adjusted). But maybe not, best way to find out is to give it a try. I won't have time for this anytime soon.
